File: frontend/src/lib/stores/sanitizer.ts
import { writable, derived } from 'svelte/store';
import { apiClient } from '../utils/api-client';

// ---- Types ----

export interface BlockedArtistBreakdown {
  artist_id: string;
  artist_name: string;
  track_count: number;
  block_reason: string;
}

export interface BlockedTrackDetail {
  track_id: string;
  track_name: string;
  artist_id: string;
  artist_name: string;
  all_artist_names: string[];
  block_reason: string;
  position: number;
  duration_ms: number;
}

export interface PlaylistGrade {
  playlist_id: string;
  playlist_name: string;
  total_tracks: number;
  clean_tracks: number;
  blocked_tracks: number;
  cleanliness_score: number;
  grade_letter: string;
  artist_breakdown: BlockedArtistBreakdown[];
  blocked_track_details: BlockedTrackDetail[];
}

export interface ReplacementTrack {
  track_id: string;
  track_name: string;
  artist_name: string;
  artist_id: string;
  album_name: string;
  popularity: number;
  preview_url: string | null;
  duration_ms: number;
  spotify_uri: string;
}

export interface ReplacementSuggestion {
  original_track_id: string;
  original_track_name: string;
  original_artist_name: string;
  candidates: ReplacementTrack[];
}

export interface PublishResult {
  new_playlist_id: string;
  new_playlist_url: string;
  tracks_kept: number;
  tracks_replaced: number;
  tracks_removed: number;
  total_tracks: number;
}

export type BatchJobStatus = 'pending' | 'grading' | 'completed' | 'failed' | 'skipped';

export interface BatchPlaylistJob {
  playlistId: string;
  playlistName: string;
  provider: string;
  status: BatchJobStatus;
  error?: string;
  gradeLetter?: string;
}

export type BatchStatus = 'running' | 'completed' | 'cancelled';

export interface BatchScrubState {
  status: BatchStatus;
  jobs: BatchPlaylistJob[];
  currentIndex: number;
}

export interface SanitizerState {
  currentGrade: PlaylistGrade | null;
  currentPlanId: string | null;
  replacements: ReplacementSuggestion[];
  selectedReplacements: Record<string, string>; // original_track_id -> replacement_track_id or "skip"
  targetPlaylistName: string;
  publishResult: PublishResult | null;
  isGrading: boolean;
  isSuggesting: boolean;
  isPublishing: boolean;
  error: string | null;
  step: 'grade' | 'replace' | 'publish';
  batchScrub: BatchScrubState | null;
}

const initialState: SanitizerState = {
  currentGrade: null,
  currentPlanId: null,
  replacements: [],
  selectedReplacements: {},
  targetPlaylistName: '',
  publishResult: null,
  isGrading: false,
  isSuggesting: false,
  isPublishing: false,
  error: null,
  step: 'grade',
  batchScrub: null,
};

// ---- Store ----

export const sanitizerStore = writable<SanitizerState>(initialState);

export const hasGrade = derived(
  sanitizerStore,
  ($s) => $s.currentGrade !== null
);

export const hasPlan = derived(
  sanitizerStore,
  ($s) => $s.currentPlanId !== null
);

export const allReplacementsSelected = derived(
  sanitizerStore,
  ($s) => {
    if ($s.replacements.length === 0) return false;
    return $s.replacements.every(
      (r) => r.original_track_id in $s.selectedReplacements
    );
  }
);

// ---- Actions ----

export const sanitizerActions = {
  /** Grade a playlist by ID or URL */
  gradePlaylist: async (playlistId: string, provider: string = 'spotify') => {
    sanitizerStore.update((s) => ({
      ...s,
      isGrading: true,
      error: null,
      currentGrade: null,
      publishResult: null,
    }));

    try {
      const response = await apiClient.post<{ grade: PlaylistGrade }>(
        '/api/v1/sanitizer/grade',
        { playlist_id: playlistId, provider }
      );

      if (response.success && response.data) {
        sanitizerStore.update((s) => ({
          ...s,
          currentGrade: response.data!.grade,
          targetPlaylistName: `${response.data!.grade.playlist_name} (Sanitized)`,
          isGrading: false,
          step: 'grade',
        }));
        return { success: true };
      } else {
        sanitizerStore.update((s) => ({
          ...s,
          error: response.message || 'Failed to grade playlist',
          isGrading: false,
        }));
        return { success: false, message: response.message };
      }
    } catch (error) {
      const msg = error instanceof Error ? error.message : 'Failed to grade playlist';
      sanitizerStore.update((s) => ({
        ...s,
        error: msg,
        isGrading: false,
      }));
      return { success: false, message: msg };
    }
  },

  /** Grade + suggest replacements, creating a draft plan */
  suggestReplacements: async (playlistId: string, provider: string = 'spotify') => {
    sanitizerStore.update((s) => ({
      ...s,
      isSuggesting: true,
      error: null,
      publishResult: null,
    }));

    try {
      const response = await apiClient.post<{
        plan_id: string | null;
        grade: PlaylistGrade;
        replacements: ReplacementSuggestion[];
        message?: string;
      }>('/api/v1/sanitizer/suggest', { playlist_id: playlistId, provider });

      if (response.success && response.data) {
        const { plan_id, grade, replacements } = response.data;

        // Auto-select first candidate for each replacement
        const autoSelected: Record<string, string> = {};
        for (const r of replacements) {
          if (r.candidates.length > 0) {
            autoSelected[r.original_track_id] = r.candidates[0].track_id;
          }
        }

        sanitizerStore.update((s) => ({
          ...s,
          currentGrade: grade,
          currentPlanId: plan_id,
          replacements,
          selectedReplacements: autoSelected,
          targetPlaylistName: `${grade.playlist_name} (Sanitized)`,
          isSuggesting: false,
          step: replacements.length > 0 ? 'replace' : 'grade',
        }));
        return { success: true };
      } else {
        sanitizerStore.update((s) => ({
          ...s,
          error: response.message || 'Failed to suggest replacements',
          isSuggesting: false,
        }));
        return { success: false, message: response.message };
      }
    } catch (error) {
      const msg = error instanceof Error ? error.message : 'Failed to suggest replacements';
      sanitizerStore.update((s) => ({
        ...s,
        error: msg,
        isSuggesting: false,
      }));
      return { success: false, message: msg };
    }
  },

  /** Select a replacement for a blocked track */
  selectReplacement: (originalTrackId: string, replacementTrackId: string) => {
    sanitizerStore.update((s) => ({
      ...s,
      selectedReplacements: {
        ...s.selectedReplacements,
        [originalTrackId]: replacementTrackId,
      },
    }));
  },

  /** Skip a blocked track (remove it instead of replacing) */
  skipTrack: (originalTrackId: string) => {
    sanitizerStore.update((s) => ({
      ...s,
      selectedReplacements: {
        ...s.selectedReplacements,
        [originalTrackId]: 'skip',
      },
    }));
  },

  /** Set the target playlist name */
  setTargetName: (name: string) => {
    sanitizerStore.update((s) => ({ ...s, targetPlaylistName: name }));
  },

  /** Confirm the plan and publish */
  confirmAndPublish: async () => {
    let planId: string | null = null;
    let selections: Record<string, string> = {};
    let targetName = '';

    sanitizerStore.update((s) => {
      planId = s.currentPlanId;
      selections = s.selectedReplacements;
      targetName = s.targetPlaylistName;
      return { ...s, isPublishing: true, error: null, step: 'publish' };
    });

    if (!planId) {
      sanitizerStore.update((s) => ({
        ...s,
        error: 'No plan to publish',
        isPublishing: false,
      }));
      return { success: false, message: 'No plan to publish' };
    }

    try {
      // Step 1: Confirm the plan
      await apiClient.put(`/api/v1/sanitizer/plan/${planId}`, {
        target_playlist_name: targetName,
        selected_replacements: selections,
      });

      // Step 2: Publish
      const response = await apiClient.post<{
        plan_id: string;
        result: PublishResult;
      }>(`/api/v1/sanitizer/publish/${planId}`);

      if (response.success && response.data) {
        sanitizerStore.update((s) => ({
          ...s,
          publishResult: response.data!.result,
          isPublishing: false,
        }));
        return { success: true };
      } else {
        sanitizerStore.update((s) => ({
          ...s,
          error: response.message || 'Failed to publish playlist',
          isPublishing: false,
        }));
        return { success: false, message: response.message };
      }
    } catch (error) {
      const msg = error instanceof Error ? error.message : 'Failed to publish playlist';
      sanitizerStore.update((s) => ({
        ...s,
        error: msg,
        isPublishing: false,
      }));
      return { success: false, message: msg };
    }
  },

  /** Navigate to a step */
  goToStep: (step: 'grade' | 'replace' | 'publish') => {
    sanitizerStore.update((s) => ({ ...s, step }));
  },

  /** Reset the sanitizer state */
  reset: () => {
    sanitizerStore.set(initialState);
  },

  /** Clear error */
  clearError: () => {
    sanitizerStore.update((s) => ({ ...s, error: null }));
  },

  // ---- Batch scrub actions ----

  /** Start a batch scrub for multiple playlists */
  startBatchScrub: async (playlists: Array<{ id: string; name?: string; playlist_name?: string; provider: string; provider_playlist_id?: string }>) => {
    const jobs: BatchPlaylistJob[] = playlists.map((p) => ({
      playlistId: p.provider_playlist_id || p.id,
      playlistName: p.name || p.playlist_name || p.id,
      provider: p.provider,
      status: 'pending' as BatchJobStatus,
    }));

    sanitizerStore.update((s) => ({
      ...s,
      batchScrub: { status: 'running', jobs, currentIndex: 0 },
      error: null,
    }));

    for (let i = 0; i < jobs.length; i++) {
      // Check if batch was cancelled
      let cancelled = false;
      sanitizerStore.update((s) => {
        if (s.batchScrub?.status === 'cancelled') cancelled = true;
        return s;
      });
      if (cancelled) break;

      // Mark current job as grading
      sanitizerStore.update((s) => ({
        ...s,
        batchScrub: s.batchScrub
          ? {
              ...s.batchScrub,
              currentIndex: i,
              jobs: s.batchScrub.jobs.map((j, idx) =>
                idx === i ? { ...j, status: 'grading' as BatchJobStatus } : j
              ),
            }
          : null,
      }));

      try {
        // Grade + suggest replacements
        const response = await apiClient.post<{
          plan_id: string | null;
          grade: PlaylistGrade;
          replacements: ReplacementSuggestion[];
        }>('/api/v1/sanitizer/suggest', {
          playlist_id: jobs[i].playlistId,
          provider: jobs[i].provider,
        });

        if (response.success && response.data) {
          const { plan_id, grade, replacements } = response.data;

          // Auto-accept first candidate for each replacement
          if (plan_id && replacements.length > 0) {
            const autoSelected: Record<string, string> = {};
            for (const r of replacements) {
              if (r.candidates.length > 0) {
                autoSelected[r.original_track_id] = r.candidates[0].track_id;
              }
            }

            await apiClient.put(`/api/v1/sanitizer/plan/${plan_id}`, {
              target_playlist_name: `${grade.playlist_name} (Sanitized)`,
              selected_replacements: autoSelected,
            });
            await apiClient.post(`/api/v1/sanitizer/publish/${plan_id}`);
          }

          sanitizerStore.update((s) => ({
            ...s,
            batchScrub: s.batchScrub
              ? {
                  ...s.batchScrub,
                  jobs: s.batchScrub.jobs.map((j, idx) =>
                    idx === i ? { ...j, status: 'completed', gradeLetter: grade.grade_letter } : j
                  ),
                }
              : null,
          }));
        } else {
          sanitizerStore.update((s) => ({
            ...s,
            batchScrub: s.batchScrub
              ? {
                  ...s.batchScrub,
                  jobs: s.batchScrub.jobs.map((j, idx) =>
                    idx === i ? { ...j, status: 'failed', error: response.message || 'Failed' } : j
                  ),
                }
              : null,
          }));
        }
      } catch (err) {
        sanitizerStore.update((s) => ({
          ...s,
          batchScrub: s.batchScrub
            ? {
                ...s.batchScrub,
                jobs: s.batchScrub.jobs.map((j, idx) =>
                  idx === i
                    ? { ...j, status: 'failed', error: err instanceof Error ? err.message : 'Unknown error' }
                    : j
                ),
              }
            : null,
        }));
      }

      // Small delay between jobs to avoid rate limits
      if (i < jobs.length - 1) {
        await new Promise((r) => setTimeout(r, 500));
      }
    }

    // Mark batch as completed (if not already cancelled)
    sanitizerStore.update((s) => ({
      ...s,
      batchScrub: s.batchScrub && s.batchScrub.status !== 'cancelled'
        ? { ...s.batchScrub, status: 'completed' }
        : s.batchScrub,
    }));
  },

  /** Cancel the current batch scrub */
  cancelBatch: () => {
    sanitizerStore.update((s) => ({
      ...s,
      batchScrub: s.batchScrub
        ? {
            ...s.batchScrub,
            status: 'cancelled',
            jobs: s.batchScrub.jobs.map((j) =>
              j.status === 'pending' ? { ...j, status: 'skipped' } : j
            ),
          }
        : null,
    }));
  },

  /** Clear batch scrub state */
  clearBatch: () => {
    sanitizerStore.update((s) => ({ ...s, batchScrub: null }));
  },
};

File: backend/crates/ndith-core/src/models/sanitizer.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::spotify::BlockReason;

/// Grade letter for playlist cleanliness
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum GradeLetter {
    A, // 90-100
    B, // 80-89
    C, // 70-79
    D, // 60-69
    F, // 0-59
}

impl GradeLetter {
    pub fn from_score(score: u32) -> Self {
        match score {
            90..=100 => GradeLetter::A,
            80..=89 => GradeLetter::B,
            70..=79 => GradeLetter::C,
            60..=69 => GradeLetter::D,
            _ => GradeLetter::F,
        }
    }
}

impl std::fmt::Display for GradeLetter {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            GradeLetter::A => write!(f, "A"),
            GradeLetter::B => write!(f, "B"),
            GradeLetter::C => write!(f, "C"),
            GradeLetter::D => write!(f, "D"),
            GradeLetter::F => write!(f, "F"),
        }
    }
}

/// Summary of blocked tracks by a single artist
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockedArtistBreakdown {
    pub artist_id: String,
    pub artist_name: String,
    pub track_count: u32,
    pub block_reason: BlockReason,
}

/// Detail about a single blocked track in a playlist
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockedTrackDetail {
    pub track_id: String,
    pub track_name: String,
    pub artist_id: String,
    pub artist_name: String,
    pub all_artist_names: Vec<String>,
    pub block_reason: BlockReason,
    pub position: u32,
    pub duration_ms: u32,
}

/// Grade result for a playlist
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlaylistGrade {
    pub playlist_id: String,
    pub playlist_name: String,
    pub total_tracks: u32,
    pub clean_tracks: u32,
    pub blocked_tracks: u32,
    pub cleanliness_score: u32,
    pub grade_letter: GradeLetter,
    pub artist_breakdown: Vec<BlockedArtistBreakdown>,
    pub blocked_track_details: Vec<BlockedTrackDetail>,
}

impl PlaylistGrade {
    pub fn compute_score(total: u32, blocked: u32) -> u32 {
        if total == 0 {
            return 100;
        }
        let clean = total.saturating_sub(blocked);
        ((clean as f64 / total as f64) * 100.0).round() as u32
    }
}

/// A candidate replacement track
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReplacementTrack {
    pub track_id: String,
    pub track_name: String,
    pub artist_name: String,
    pub artist_id: String,
    pub album_name: String,
    pub popularity: u32,
    pub preview_url: Option<String>,
    pub duration_ms: u32,
    pub spotify_uri: String,
}

/// Replacement suggestion for a single blocked track
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReplacementSuggestion {
    pub original_track_id: String,
    pub original_track_name: String,
    pub original_artist_name: String,
    pub candidates: Vec<ReplacementTrack>,
}

/// Status of a sanitization plan
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum SanitizationStatus {
    Draft,
    Confirmed,
    Publishing,
    Published,
    Failed,
}

impl std::fmt::Display for SanitizationStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SanitizationStatus::Draft => write!(f, "draft"),
            SanitizationStatus::Confirmed => write!(f, "confirmed"),
            SanitizationStatus::Publishing => write!(f, "publishing"),
            SanitizationStatus::Published => write!(f, "published"),
            SanitizationStatus::Failed => write!(f, "failed"),
        }
    }
}

impl SanitizationStatus {
    pub fn from_str(s: &str) -> Self {
        match s {
            "draft" => SanitizationStatus::Draft,
            "confirmed" => SanitizationStatus::Confirmed,
            "publishing" => SanitizationStatus::Publishing,
            "published" => SanitizationStatus::Published,
            "failed" => SanitizationStatus::Failed,
            _ => SanitizationStatus::Draft,
        }
    }
}

/// Full sanitization plan persisted to DB
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SanitizationPlan {
    pub id: Uuid,
    pub user_id: Uuid,
    pub provider: String,
    pub source_playlist_id: String,
    pub source_playlist_name: String,
    pub target_playlist_name: Option<String>,
    pub grade: PlaylistGrade,
    pub replacements: Option<Vec<ReplacementSuggestion>>,
    pub selected_replacements: Option<std::collections::HashMap<String, String>>,
    pub publish_result: Option<PublishResult>,
    pub status: SanitizationStatus,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub published_at: Option<DateTime<Utc>>,
}

/// Result of publishing a sanitized playlist
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PublishResult {
    pub new_playlist_id: String,
    pub new_playlist_url: String,
    pub tracks_kept: u32,
    pub tracks_replaced: u32,
    pub tracks_removed: u32,
    pub total_tracks: u32,
}

/// Request to grade a playlist
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GradePlaylistRequest {
    pub playlist_id: String,
}

/// Request to suggest replacements (grade + suggest)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SuggestReplacementsRequest {
    pub playlist_id: String,
}

/// Request to confirm a plan with user selections
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfirmPlanRequest {
    pub target_playlist_name: String,
    /// Map of original_track_id -> chosen replacement_track_id (or "skip" to drop)
    pub selected_replacements: std::collections::HashMap<String, String>,
}

/// Response from grading
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GradeResponse {
    pub grade: PlaylistGrade,
}

/// Response from suggesting replacements (creates a draft plan)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SuggestResponse {
    pub plan_id: Uuid,
    pub grade: PlaylistGrade,
    pub replacements: Vec<ReplacementSuggestion>,
}

/// Response from publishing
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PublishResponse {
    pub plan_id: Uuid,
    pub result: PublishResult,
}
